### PR TITLE
Feat: add xls export method

### PR DESCRIPTION
**Pull request by NicolasRichel**

### CODE REVIEW

File: bcf_api_xml/export.py [hunk 1]
```diff
@@ -84,7 +89,93 @@ def to_zip(topics, comments, viewpoints):
...
+    headerFmt = workbook.add_format({ "align": "center", "bold": True, "bg_color": "#C0C0C0" })
+    baseFmt = workbook.add_format({ "valign": "top" })
+    dateFmt = workbook.add_format({ "valign": "top", "num_format": "yyyy-mm-dd" })
+
+    row = 0
+
+    # TODO: add spreadsheet metadata
+
+    # Create table header
+    worksheet.write(row, 0, "N°",              headerFmt)
```
Comment (Amoki) [inline bcf_api_xml/export.py:117]: i18n

File: bcf_api_xml/export.py [hunk 2]
```diff
@@ -84,7 +89,93 @@ def to_zip(topics, comments, viewpoints):
...
+def to_xls(topics, comments, viewpoints):
+    """
+    topics: list of topics (dict parsed from BCF-API json)
+    comments: dict(topics_guid=[comment])
+    viewpoints: dict(topics_guid=[viewpoint])
+    """
+    workbook = xlsxwriter.Workbook("bcf-export.xlsx")
+    worksheet = workbook.add_worksheet()
+
+    headerFmt = workbook.add_format({ "align": "center", "bold": True, "bg_color": "#C0C0C0" })
```
Comment (Amoki) [inline bcf_api_xml/export.py:108]: snake_case

File: bcf_api_xml/export.py [hunk 3]
```diff
@@ -84,7 +89,93 @@ def to_zip(topics, comments, viewpoints):
...
+                img_file = "snapshot-"+viewpoint["guid"]+".png"
+                if ";base64," in snapshot:
+                    _, img_data = snapshot.split(";base64,")
+                    img_data = base64.b64decode(img_data)
+                else:
+                    img_data = requests.get(snapshot).content
+
+                with open(img_file, "wb") as f:
+                    f.write(img_data)
+                with Image.open(img_file) as img:
```
Comment (Amoki) [inline bcf_api_xml/export.py:167]: use BytesIO

File: bcf_api_xml/export.py [hunk 4]
```diff
@@ -84,7 +89,93 @@ def to_zip(topics, comments, viewpoints):
...
+                    width, height = img.size
+                scale = 300 / width
+
+                worksheet.set_column_pixels(9, 9, 300)
+                worksheet.set_row_pixels(row, height * scale)
+                worksheet.insert_image(row, 9, img_file, { "x_scale": scale, "y_scale": scale })
```
Comment (Amoki) [inline bcf_api_xml/export.py:173]: use image_data (https://xlsxwriter.readthedocs.io/worksheet.html#worksheet-insert-image)

File: bcf_api_xml/models/Topic.py [hunk 5]
```diff
@@ -59,17 +58,11 @@ def to_python(xml):
     if (index := xml.find("Index")) is not None and index.text is not None:
         topic["index"] = index.text
 
-    if (creation_date := xml.find("CreationDate")) is not None and creation_date.text is not None:
-        topic["creation_date"] = parse(creation_date.text)
```
Comment (Amoki) [inline bcf_api_xml/models/Topic.py:63]: keep the parse

**Review by Amoki — CHANGES_REQUESTED** [review]

File: tests/data/topics.json [hunk 6]
```diff
@@ -1,16 +1,191 @@
...
+    "labels":[],
+    "creation_date":"2021-09-30T06:36:18.186819Z",
+    "creation_author":"gaetan+demo@bimdata.io",
+    "modified_date":"2021-09-30T06:36:18.187917Z",
+    "modified_author":"gaetan+demo@bimdata.io",
+    "assigned_to":"gaetan@bimdata.io",
+    "stage":"Constructioneevv",
+    "description":"Les rambardes sont pourries...",
+    "due_date":"2021-12-25T07:34:00Z",
+    "ifcs":[],
```
Comment (Amoki) [inline tests/data/topics.json:140]: remove non standard fields

File: bcf_api_xml/export.py [hunk 7]
```diff
@@ -84,7 +89,93 @@ def to_zip(topics, comments, viewpoints):
...
+            due_date = datetime.strptime(due_date, "%Y-%m-%dT%H:%M:%SZ")
+            worksheet.write_datetime(row, 6, due_date, dateFmt)
+
+        worksheet.write(row, 7, topic.get("topic_status"), baseFmt)
+        worksheet.write(row, 8, topic.get("priority"), baseFmt)
+
+        for viewpoint in topic_viewpoints:
+            if viewpoint.get("snapshot"):
+                snapshot = viewpoint.get("snapshot").get("snapshot_data")
+                img_file = "snapshot-"+viewpoint["guid"]+".png"
```
Comment (Amoki) [inline bcf_api_xml/export.py:158]: f-string

**Review by Amoki — CHANGES_REQUESTED** [review]

File: bcf_api_xml/export.py [hunk 8]
```diff
@@ -96,86 +123,90 @@ def to_zip(topics, comments, viewpoints):
...
+    xls_file = io.BytesIO()
+    with xlsxwriter.Workbook(xls_file) as workbook:
+        worksheet = workbook.add_worksheet()
+
+        header_fmt = workbook.add_format({ "align": "center", "bold": True, "bg_color": "#C0C0C0" })
+        base_fmt = workbook.add_format({ "valign": "top" })
+        date_fmt = workbook.add_format({ "valign": "top", "num_format": "yyyy-mm-dd" })
+        comments_fmt = workbook.add_format({ "valign": "top", "text_wrap": True })
+
+        headers = XLS_HEADER_TRANSLATIONS[lang]
```
Comment (Amoki) [inline bcf_api_xml/export.py:141]: is lang does not exist, fallback to en

File: bcf_api_xml/export.py [hunk 9]
```diff
@@ -96,86 +123,90 @@ def to_zip(topics, comments, viewpoints):
...
+        # Create topic rows
+        for topic in topics:
+            topic_guid = topic["guid"]
+            topic_comments = comments.get(topic_guid, [])
+            topic_viewpoints = viewpoints.get(topic_guid, [])
+
+            worksheet.write(row, 0, topic.get("index"), base_fmt)
+            creation_date = topic.get("creation_date")
+            if creation_date:
+                if isinstance(creation_date, str):
```
Comment (Amoki) [inline bcf_api_xml/export.py:170]: it will always be a string as the input is the standardized json, no?

**Review by Amoki — CHANGES_REQUESTED** [review]

File: bcf_api_xml/export.py (hunk 8 above)
Comment (NicolasRichel) [reply bcf_api_xml/export.py:141]: I already defined `"en"` as [the default value for `lang` here](https://github.com/bimdata/BCF-API-XML-translator/pull/9/files#diff-d04fee44370a0aa1f24bb8bdc75ae6919c6e4d6730fde99542ed73156ba99be7R126).
Is this OK or should I do it another way ?

**Pull request merged**